# Primeros pasos con Python

## 1. Ejecutando  Python

Uno de los desafíos al comenzar a trabajar con Python es instalar el software relacionado en tu computadora.

Si estás familiarizado con su sistema operativo, y especialmente si te sentís cómodo con la interfaz de línea de comandos, no vas a tener problemas para instalar Python. Pero para los principiantes, puede ser complicado  aprender sobre administración del sistema y programación al  mismo tiempo.

Para evitar ese problema, recomendamos comenzar con Jupyter Notebooks. *Jupyter Notebook* (anteriormente IPython Notebooks) es un entorno informático que permite trabajar alternadamente con programas Python y texto enriquecido. Este que están viendo es un *cuaderno Jupyter*.  

Existen varias implementaciones de cuadernos Jupyter para comezar a trabajar con ellas:

- *Google Colaboratory (Colab).* Es una versión de Jupyter Notebooks on line provista por Google. Es una fabulosa herramienta didáctica y nos permitirá hacer gran parte del curso aunque en  las últimas semanas posiblemente necesitemos otra herramienta. Las tareas y ejercicios del curso se harán con esta herramienta.
- *Anaconda.* Anaconda es mucho más que Jupyter Notebooks y los que se sientan cómodos con esta aplicación podrán utilizarla en el curso.
- *JupyterLab Desktop.*  Es un subproducto de Anaconda, pero mucho más fácil de instalar y manejar.

Un  cuaderno Jupyter se divide en celdas de texto con formato y celdas de código Python. Cuando creamos una celda elegimos que tipo de celda queremos ("Markdown" para texto, "Code" para Python). Obviamente lo que estamos leyendo ahora es una celda de texto. La celda siguiente es una celda de código y la podemos ejecutar cliqueando el triangulito o posicionarse en ella  y hacer `Mayus+Enter`:




In [ ]:
1 + 1

2

En muchos proyectos que involucren desarrollo de software en Python se utiliza *el modo de secuencia de comandos* o *modo script*. En este modo cada pieza de código se guarda en un archivo y luego se ejecuta este archivo, llamado *script*, en el intérprete. Por convención, los scripts de Python tienen nombres que terminan en `.py`. Si sabés cómo crear y ejecutar un script en tu computadora podrás correr todos los scripts de este curso.

Hay herramientas muy útiles para escribir código llamadas IDE (Integrated Development Environment) o entornos de desarrollo integrado​s. Una IDE es una aplicación informática que proporciona servicios integrales para facilitarle al desarrollador o programador el desarrollo de software. 

La primera que mencionaremos es Visual Studio Code. Visual Studio Code es un editor de código fuente desarrollado por Microsoft para Windows, Linux, macOS y Web. Incluye soporte para la depuración, control integrado de Git, resaltado de sintaxis, finalización inteligente de código, fragmentos y refactorización de código. También es personalizable, por lo que los usuarios pueden cambiar el tema del editor, los atajos de teclado y las preferencias. Es gratuito y de código abierto. Por medio de plugins es posible codificar en casi cualquier lenguaje de programación, en particular en Python. 

La otra herramienta a mencionar es PyCharm. PyCharm es un IDE multiplataforma dirigido principalmente a Python y al desarrollo web. Una versión de código abierto está disponible como PyCharm Community Edition. Hay una versión paga llamada PyCharm Professional Edition. Una versión especial llamada PyCharm Edu está basada en PyCharm, dirigida específicamente al aprendizaje de la programación con Python. Como alumno universitario podés aplicar a la versión académica. PyCharm tiene una distribución para Anaconda que se puede instalar junto con Anaconda y están estrechamente integradas.

En la primer celda que ejecutamos usamos a Python como calculadora y ese es el uso  mas sencillo que le podemos dar. Profundizaremos más abajo en este aspecto. 

Ahora vemos algunos ejemplos muy sencillos de scripts:

In [ ]:
print('¡Hola, mundo!') 

¡Hola, mundo!


Este es un ejemplo de una _sentencia print_, aunque en realidad no imprime nada en papel. Muestra un resultado en la pantalla. En este caso, como ya habrás comprobado, el resultado son las palabras
```
¡Hola Mundo!
```
Las comillas en el programa,  que pueden ser simples o dobles, marcan el principio y el final del texto que se mostrará y no aparecen en el resultado.

Los paréntesis indican que `print` es una _función_. Veremos funciones con profundidad más adelante.


In [ ]:
print('El "Tata" Martino')
print("El 'Tata' Martino")
print("El \"Tata\" Martino")
print("El \"Tata\" Martino\nEntrenador")


El "Tata" Martino
El 'Tata' Martino
El "Tata" Martino
El "Tata" Martino
Entrenador


## 2. Operadores aritméticos

Después de `Hola, mundo`, el siguiente paso es la aritmética elemental. Python proporciona _operadores_, que son símbolos especiales que representan cálculos
como suma y multiplicación.

Los operadores `+`, `-` y `*` realizan sumas, resta y multiplicación, respecivamente. Por ejemplo:


In [ ]:
40 + 2

42

In [ ]:
43 - 1

42

In [ ]:
6 * 3

18

El operador `/` realiza la división:



In [ ]:
84 / 2

42.0

Quizás te preguntes al ejecutar la celda anterior por qué el resultado es `42.0` en lugar de `42`. La respuesta es que Python tiene dos operadores de división. el primero,  es el que ya vimos, es `/` que siempre obtiene  un número decimal o coma flotante. El segundo es `//` y es la _división entera_. Por ejemplo, 


In [ ]:
84 // 2

42

devuelve `42` y 

In [ ]:
20 // 3

6

Devuelve `6`, pues "3 entra 6 veces en 20".

Finalmente, el operador `**` realiza una exponenciación; esto es,
eleva un número a una potencia:

In [ ]:
6 ** 2 

36

Sabemos que $a^n$ es $a$ multiplicado por si mismo $n$-veces. También conocemos, que $(a^n)^m = a^{nm}$. 

Enntonces, como sabemos que $(\sqrt[n]{a})^n = a$, es "claro" que $\sqrt[n]{a} = a^{1/n}$, pues $ (a^{1/n})^n = a^{\frac1n n} =a$.

Recordemos que  $\sqrt[n]{a} = a^{1/n}$. Luego
$$
\sqrt{10} = 10^{1/2} = 10^{0.5}.
$$
Por ejemplo, 

In [ ]:
print(9**0.5) # raíz cuadrada de 9 (= 3)
print(27**(1/3)) # raíz cúbica de 27 (= 3)

3.0
3.0


En  las celdas de código anteriores usamos el símbolo `#` para comenzar un *comentario*: el texto que sigue después de `#` no será interpretado  o ejecutado y sirve de ayuda al programador.

En algunos lenguajes de programación se utiliza `^` para la exponenciación, pero en Python `^` es un operador bit a bit llamado XOR. Si no estás familiarizado con los operadores bit a bit, el resultado te sorprenderá:



In [ ]:
6 ^ 2 

4

No cubriremos los operadores bit a bit en este curso, pero podes leer acerca de
ellos en [http://wiki.python.org/moin/BitwiseOperators](http://wiki.python.org/moin/BitwiseOperators) (en inglés). 

## 3. Valores y tipos

Un _valor_ es una de las entidades básicas con las que trabaja un programa, como una letra o un número. Algunos valores que hemos visto hasta ahora son `2`, `42.0` y `'¡Hola, mundo!'`.

Estos valores pertenecen a diferentes _tipos_: `2` es un _entero_, `42.0` es un _número de coma flotante_ y `'¡Hola, mundo!'`es una _cadena_, así llamada porque los caracteres que contiene se ubican en forma consecutiva.

Si no estás seguro de qué tipo de valor es un valor determinado, Python mismo puede decírtelo ejecutando la función `type()`.


In [ ]:
print(type(2))
print(type(42.0))
print(type('¡Hola, mundo!'))
print(type(True))

<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>


In [ ]:
n = 2
print( n < 5)

True


Vemos en los resultados de ejecutar la celda anterior que en todos aparece la palabra `class`. La palabra `class`,  cuya traducción al español es *clase*, se usa en el sentido de una categoría: un tipo es una categoría de valores.

No es sorprendente que los enteros pertenezcan al tipo `int`, las cadenas pertenezcan a `str` (por _string_ en inglés) y los números en coma flotante pertenezcan a `float`.

El tipo `bool` solo tiene doa valores `True` y `False`,  que sirven para denotar la verdad o falsedad de expresiones lógicas.

¿Qué pasa con valores como `'2'` y `'42.0'`?
Parecen números, pero están entre comillas como cadenas y, efectivamente son cadenas. Lo  comprobaremos ejecutando lo siguiente:


In [ ]:
print(type('2'))
print(type('42.0'))

<class 'str'>
<class 'str'>


Cuando escribís un entero grande, podés tener la tentación de usar puntos entre grupos de dígitos, como en `1.000.000`. Sin embargo,  esto no es una forma permitida de escribir un entero, por ejemplo, la ejecución de la siguiente celda de código nos devuelve un error:

In [ ]:
print(1.000.000)

La forma correcta de escribir "un millón" es `1000000`. Podemos tratar también de escribir un número decimal o coma flotante usando la notación latina. Por ejemplo, "cuatro coma cinco", podríamos intentar escribirlo como `4,5`,  y hagamos la prueba:  

In [ ]:
print(4,5)

4 5


¡Eso no es lo que esperábamos en absoluto! Python interpreta la expresión `4,5` como dos enteros separados. 

Este es el primer ejemplo que hemos visto de un *error semántico*: el código se ejecuta sin producir un mensaje de error, pero no es lo que queremos obtener.

La forma correcta de escribir un número decimal es reemplazando  la coma por  el punto, por ejemplo "cuatro coma cinco" se escribe `4.5`

In [ ]:
print(4.5 * 2)

## 4. Python como calculadora 

Como vimos más arriba Python puede ser usado como una calculadora. Ahora veremos más funciones de "calculadora". Para ello utilizaremos el módulo `math`. El módulo `math` permite utilizar funciones matemáticas que no son accesibles directamente en Python. La instrucción para importar el módulo es 

In [ ]:
import math

y debemos ejecutarla si en las siguientes celdas de código queremos usar el módulo. 

La instrucción para obtener el logaritmo en base $e$ es `math.log(t)`, donde `t` es un número real positivo. 

In [ ]:
math.log(10)

2.302585092994046

Si queremos obtener el logaritmo en base 10 debemos usar la función `math.log10()`:

In [ ]:
print(math.log10(10))
print(math.log10(10**234))

1.0
234.0


Las funciones trigonométricas también son accesibles a travez del módulo `math`.

In [ ]:
math.sin(0)

0.0

es el seno  de $0$. 

Si queremos calcular el seno de 360° directamente fallaremos:

In [ ]:
math.sin(360)

0.9589157234143065

El problema es que las funciones trigonométricas en Python aceptan  los grados en radianes.  Recordemos que $\pi$ radianes son 180° sexagesimales, todo lo demás es regla de tres simple:
- 360° $\to$ $2\pi$
- 270° $\to$ $\frac32 \pi$
- 90° $\to$ $\pi/2$

En general
- $x$ grados sexagesimales $\to$ $x \cdot \pi / 180$ radianes.

Podemos obtener $\pi$ copiando alguna estimación de algún sitio web, pero es más sencillo utilizar el que nos provee  Python: 

In [ ]:
math.pi

3.141592653589793

In [ ]:
math.sin(2 * math.pi)

-2.4492935982947064e-16

La expresión `-2.4492935982947064e-16` se lee $-2.4492935982947064 \times 10^{-16}$. Es decir es
$$
-0.00000000000000024492935982947064
$$
Nosotros esperaríamos que el resultado de ejecutar la línea de código anterior sea `0`, pero debido a que el valor de $\pi$ es una aproximación y  que la función `sin()` también devuelve aproximaciones,  el resultado es número "cercano"  a `0`.

Para ver todas las funciones del módulo ejecutamos `dir(math)`:

In [ ]:
dir(math)

['__doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'acos',
 'acosh',
 'asin',
 'asinh',
 'atan',
 'atan2',
 'atanh',
 'ceil',
 'copysign',
 'cos',
 'cosh',
 'degrees',
 'e',
 'erf',
 'erfc',
 'exp',
 'expm1',
 'fabs',
 'factorial',
 'floor',
 'fmod',
 'frexp',
 'fsum',
 'gamma',
 'gcd',
 'hypot',
 'inf',
 'isclose',
 'isfinite',
 'isinf',
 'isnan',
 'ldexp',
 'lgamma',
 'log',
 'log10',
 'log1p',
 'log2',
 'modf',
 'nan',
 'pi',
 'pow',
 'radians',
 'remainder',
 'sin',
 'sinh',
 'sqrt',
 'tan',
 'tanh',
 'tau',
 'trunc']

Entre todos los métodos que nos provee la biblioteca `math` tenemos los que convierten radianes a grados sexagesimales y  viceversa.

In [ ]:
math.degrees(math.pi)

180.0

In [ ]:
math.sin(math.radians(360)) # esto debería resultar un número cercano a 0

-2.4492935982947064e-16

In [ ]:
print(math.e)
math.log(math.e)

2.718281828459045


1.0

Recordemos que $\log(x) = y$ si y solo si $e^y = x$.

En  general 
$$
\log_a(x) = y \quad\Leftrightarrow \quad a^y = x  \quad \Leftrightarrow\quad  \log(a^y) = \log(x) \quad  \Leftrightarrow \quad \log(a) \cdot y = \log(x)\quad  \Leftrightarrow \quad  y = \log(x) / \log(a)
$$

Concluyendo:
$$
\log_a(x) =  \log(x) / \log(a)
$$

Luego  si queremos calcular, el logaritmo en base 2 de `x`, debemos  hacer `math.log(x) / math.log(2)`. Por ejemplo,

In [ ]:
print(math.log(10) / math.log(2)) # logartimo en base 2 de 10
print(math.log(16) / math.log(2)) # logartimo en base 2 de 16 (es 4, pues 2**4 = 16)
print(math.log2(10))
print(math.log2(16))

3.3219280948873626
4.0
3.321928094887362
4.0


Otra forma de elevar un número  a una potencia es utilizando `math.pow()`

In [ ]:
math.pow(2, 3)

8.0

que es lo mismo que

In [ ]:
2**3

8

También, para ciertas situaciones y algoritmos a veces es conveniente manejar el concepto de infinito. Esto muchas veces se hace eligiendo un número "muy grande". Python nos provee `math.inf` que pese a no ser un númerpo formalmente es del tipo `float`. Decimos que no es número porque no cumple las leyes o axiomas de un número real pues es un "número" más grande que cualquier otro. 

Para que quede claro: matemáticamente *infinito* no es un número sino un concepto matemático.

In [ ]:
type(math.inf)

float

Por  el tipo `math.inf` parece ser un número, pero...

In [ ]:
math.inf + 1 == math.inf # ningún número cumple con esta condición
math.inf / 2 == math.inf

0.0

El `==` compara dos cantidades y devuelve `True` si son iguales.

In [ ]:
print(10**452 < math.inf) # todos los números cumplen con esta condición

True


Para evitar escribir `math.` antes de cada función ejecutamos la siguiente celda  de código:

In [ ]:
from math import *


Luego de ejecutar la celda de arriba es lícito escribir, por ejemplo:

In [ ]:
sin(2 * pi)

-2.4492935982947064e-16

Si sabemos que vamos a utilizar algunas funciones determinadas del módulo, podemos importar solamente esa funciones:

In [ ]:
from math import log, e
log(e)

1.0

Si queremos acortar o cambiar el nombre del módulo dentro de nuestro programa podemos hacer algo como lo siguiente:

In [ ]:
import math as mt
mt.log(mt.e)